### PR TITLE
V4/slack integration

### PR DESCRIPTION
**Pull request by sinmi-hub**
#### Asana Link: https://app.asana.com/1/8442528107068/project/1209571477232011/task/1211203223024106?focus=true

---

## Detailed AC

- [x] Ensure a wrapper file exists for slack under assets/wrappers/python folder
```**slack_bolt_wrapper file.py currently exists under asset/wrappers/python**```

- [x] Ensure that the wrapper file has a client method or similar for creating a SlackClient. This may require pulling in the slack bot token from the SSM parameter store to authenticate the client for performing operations in Slack.
      ```**Authentication is handled with SlackBoltClient class within _create_app() method. SSM Parameter Store integration is present through the _get_bot_token() and  _get_signing_secret() method**```
       
- [x] Ensure that the wrapper file has other functions or methods to perform Slack operations like create_slack_channel, get_channel_details, post_message, upload_attachment etc. 
         ```**The functions above are present and use self.client properly**```
         
- [x] These methods should use the SlackClient instantiated for the class. 

- [x] Using this client, ensure that each function is using the appropriate Slack API based on the operation being performed.

     ```
     Correct slack api such as:
          -  conversations_create for channel creation
          - conversations_setTopic for setting topics
          - chat_postMessage for messaging
          - files_upload_v2 for file uploads
          - conversations_info for channel details

- [x] Ensure the right parameters (required or optional) are being passed to the Slack API calls (See https://docs.slack.dev/tools/bolt-python/ documentation for more details.)
    ```
      -      Correct Slack APIs (conversations_create, chat_postMessage, etc.)
      -      Proper required/optional parameters(name, channel, isPrivate, etc, as applicable)
      -      Follows Slack Bolt documentation patterns 
    ```
- [x] Add any missing operation based on the requirement details in .kiro/specs/slack-integration/requirements.md
  ``` There was no missing operation. However, operations that did not relate to wrappers were not validated, since the current task focused on wrapper validation```

---
**CLARIFICATION REQUIRED HERE:**
 Requirement 3 in Kiro states: 
 _WHEN case watchers are added or removed THEN the system SHALL add or remove users from the Slack channel accordingly_
 
 Does this mean a remove method has to be implemented? maybe removing users from channel? Is this a logic to be implemented such that case watcher that is added to the slack channel should automatically be removed after? if so, this method is missing and would require implementation

---

- [x] Using Slack Client, ensure that the newly added function is using the appropriate Slack API based on the operation being performed.

- [x] Ensure that the right parameters (required or optional) are being passed to the respective Slack API calls

- [x] Ensure there is proper error handling via try, except blocks for each method
```
- Try/except blocks in all methods
-  _retry_with_backoff() with exponential backoff
-  SlackApiError handling
```

- [x] Ensure all the appropriate imports exist including the domain models
```some imports had to be added (given that the tests failed automatically)```

- [x] Ensure there is logging setup with appropriate/necessary log statements having structured logging format

- [x] Ensure that the wrapper file has proper class definitions and structure

- [x] Ensure Pydocs exist for all the domain models methods
```pydoc strings were added to the methods```

- [x] Ensure Type hints for all attributes and methods in the domain model
```Pylance was activated within the virtual environment and set to the strictest standard, which revealed a lot of type annotation errors which was fixed```

- [x] Ensure tests for slack wrapper exists under tests/assets/wrappers/ 

- [x] Ensure mock exists for Slack Client, all Slack API calls using unittest.mock or pytest-mock

- [x] Ensure that the tests validate successful API responses and error scenarios
---
**ATTENTION REQUIRED HERE:**
In regards to this task, **there were issues**. One of the test were broken due to how mocking was done in the test. This is the test_message_sync.py. Given that I am not familiar with test mocking in python, and Q struggled quite hard to fix it.

 I had to refactor it into test_message_sync.disabled, since I did not want to delete it. However, given that the functions tested in message_sync were used in other tests, and those tests were passing, I decided to proceed with the rest of the task

---

- [x] Ensure tests exists for all wrapper methods and properties (Optionally achieve >90% code coverage including error paths)
```Test currently achieve 95% coverage(due to disabled test above). It has been verified to be more of a configuration issue in the tests, than functional issues in the code.```

- [x] Ensure validation of authentication token retrieval from SSM. Test authentication failure: mock SlackApiError with error='invalid_auth'

- [x] Test channel creation with various parameters and edge cases

- [x] Verify error categorization and logging output

- [x] Test file upload with different file types and sizes


## Code Updates

- fix: change use of deprecated function in  datetime.utcnow() → datetime.now(datetime.timezone.utc). There is multiple use of datetime.utcnow() across source files and test files. This has been deprecated from Python 3.12+. Could cause issues

- Removed Code Duplications for the following in assets/slack_client/index.py :
    - Removed duplicate get_case() method (lines 175-197)
    - Removed inferior _get_slack_user_name() method (lines 727-749)
    - Fixed return type consistency in remaining get_case() method
    - Fixed data access pattern in sync_watchers_to_slack() method
    - Fixed return type mismatch in process_case_event() method
Ran tests again after these changes and all 33 tests were passing

### CODE REVIEW

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 1]
```diff
@@ -203,24 +249,37 @@ def _create_channel():
...
         """
         if not self.client:
             logger.error("Slack client not initialized")
             return False
 
         try:
-            def _post_message():
-                return self.client.chat_postMessage(
+            def _post_message() -> Any:
+                return self.client.chat_postMessage(  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:282]: issue (non-blocking): should we really choose to ignore type errors? I'd recommend removing the `type: ignore` unless its absolutely necessary

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 2]
```diff
@@ -247,8 +306,8 @@ def add_users_to_channel(self, channel_id: str, user_ids: List[str]) -> bool:
             return False
 
         try:
-            def _invite_users():
-                return self.client.conversations_invite(
+            def _invite_users() -> Any:
+                return self.client.conversations_invite(  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:310]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 3]
```diff
@@ -274,9 +333,9 @@ def get_channel_info(self, channel_id: str) -> Optional[Dict[str, Any]]:
             return None
 
         try:
-            def _get_channel_info():
-                response = self.client.conversations_info(channel=channel_id)
-                return response["channel"]
+            def _get_channel_info() -> Any:
+                response = self.client.conversations_info(channel=channel_id)  # type: ignore
+                return response["channel"]  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:338]: issue (non-blocking): please see previous comment. Excessive use of `type: ignore` can end up resulting into unacknowledged or trickle down into further exceptions in the calling functions.

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 4]
```diff
@@ -285,26 +344,39 @@ def _get_channel_info():
...
         """
         if not self.client:
             logger.error("Slack client not initialized")
             return False
 
         try:
-            def _upload_file():
-                return self.client.files_upload_v2(
+            def _upload_file() -> Any:
+                return self.client.files_upload_v2(  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:379]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 5]
```diff
@@ -333,10 +405,10 @@ def remove_users_from_channel(self, channel_id: str, user_ids: List[str]) -> boo
             return False
 
         try:
-            def _kick_users():
+            def _kick_users() -> Dict[str, bool]:
                 # Slack API requires individual kick operations
                 for user_id in user_ids:
-                    self.client.conversations_kick(
+                    self.client.conversations_kick(  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:411]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 6]
```diff
@@ -364,8 +436,8 @@ def update_channel_topic(self, channel_id: str, topic: str) -> bool:
             return False
 
         try:
-            def _set_topic():
-                return self.client.conversations_setTopic(
+            def _set_topic() -> Any:
+                return self.client.conversations_setTopic(  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:440]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 7]
```diff
@@ -391,9 +463,9 @@ def get_channel_members(self, channel_id: str) -> Optional[List[str]]:
             return None
 
         try:
-            def _get_members():
-                response = self.client.conversations_members(channel=channel_id)
-                return response["members"]
+            def _get_members() -> List[str]:
+                response = self.client.conversations_members(channel=channel_id)  # type: ignore
+                return response["members"]  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:468]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 8]
```diff
@@ -415,8 +487,8 @@ def archive_channel(self, channel_id: str) -> bool:
             return False
 
         try:
-            def _archive_channel():
-                return self.client.conversations_archive(channel=channel_id)
+            def _archive_channel() -> Any:
+                return self.client.conversations_archive(channel=channel_id)  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:491]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 9]
```diff
@@ -429,27 +501,37 @@ def get_user_info(self, user_id: str) -> Optional[Dict[str, Any]]:
...
             logger.error("Slack client not initialized")
             return None
 
         try:
-            def _get_user_info():
-                response = self.client.users_info(user=user_id)
-                return response["user"]
+            def _get_user_info() -> Dict[str, Any]:
+                response = self.client.users_info(user=user_id)  # type: ignore
+                return response["user"]  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:526]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 10]
```diff
@@ -467,14 +549,14 @@ def get_channel_messages(self, channel_id: str, oldest: str = None, latest: str
             return None
 
         try:
-            def _get_messages():
-                response = self.client.conversations_history(
+            def _get_messages() -> List[Dict[str, Any]]:
+                response = self.client.conversations_history(  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:553]: issue (non-blocking): please see previous comment

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 11]
```diff
@@ -467,14 +549,14 @@ def get_channel_messages(self, channel_id: str, oldest: str = None, latest: str
...
                     channel=channel_id,
                     oldest=oldest,
                     latest=latest,
                     limit=limit
                 )
-                return response["messages"]
+                return response["messages"]  # type: ignore
```
Comment (nshalabh) [inline assets/wrappers/python/slack_bolt_wrapper.py:559]: issue (non-blocking): please see previous comment

File: tests/assets/slack_client/test_message_sync.py.disabled
Comment (nshalabh) [inline tests/assets/slack_client/test_message_sync.py.disabled:1]: recommendation: can you please add a `TODO` saying `Temporarily disabled due to import conflicts`. This needs to be resolved before we can merge to `main`. I can help with that.

**Review by nshalabh — CHANGES_REQUESTED** [review]
@sinmi-hub Thank you for publishing the PR. Can you please look at the comments, make updates accordingly?

Overall, couple more notes OR issues (blocking):
1. Please see that the title of the PR matches the title of the workitem that is related to the PR. Eg. in this case, it should be `Validate Slack API interface and wrappers implementation`
2. Somehow, the PR did not trigger the required Checks from the GitHub (tests, security scans etc.), which are defined in `.github/workflows/security-and-tests.yml`. We cannot merge without completing these checks, eg PR - https://github.com/aws-samples/sample-aws-security-incident-response-integrations/pull/27

File: assets/wrappers/python/slack_bolt_wrapper.py [hunk 12]
```diff
@@ -118,7 +146,26 @@ def _get_signing_secret(self) -> Optional[str]:
...
+            delay: Current delay value
+            error_msg: Error message to log
+            
+        Returns:
+            Updated delay value
+        """
+        if attempt < SLACK_MAX_RETRIES - 1:
+            logger.warning(f"{error_msg} on attempt {attempt + 1}")
+            time.sleep(delay)
+            return min(delay * 2, SLACK_MAX_RETRY_DELAY)
```
[the remark this answers is not in the document]
Comment (sinmi-hub) [reply assets/wrappers/python/slack_bolt_wrapper.py:163]: This is a very vital suggestion that I have missed. Lambda timeout is not taken into consideration.

The new approach is to check lambda timeout at the end of each iteration to see if its closer to timeout 15 minutes. A 30 second buffer is used for computation (so 14minutes and  30seconds, to be accurate). The major change here is to implement retry logic to be aware of the lambda time. 

**issue(blocking)**: The exception logic retries after a significant amount of delay. However, this might be too broad. The issue here is _not failing fast_, as exceptions like AttributeError or TypeError could be caught, which would require manual fix. Keeping a retry mechanism in this case would waste resources and cause lambda to reach its timeout fast. Replaced broad exception handling with specific network error retry logic (ConnectionError, TimeoutError, OSError) that actually would benefit from a try. 

I can also simply add a TODO to be fixed for later if needed

File: assets/wrappers/python/slack_bolt_wrapper.py (hunk 1 above)
Comment (sinmi-hub) [reply assets/wrappers/python/slack_bolt_wrapper.py:282]: Apologies for the excessive use of type:ignore, but this is due to pylance type checking which caused numerous possible type errors for me locally, but did not affect functionality. After verifying that it was not an issue(in context of current code only), I added those.

**issue(non-blocking)**: It could be good to add pylance as a requirement for this project, as it could assist current team members in understand and finding possible type errors, a lot faster. While all type errors might not affect functionality, it provides a faster and easier way to debug type errors or even logical ones that might exist

File: tests/assets/slack_client/test_message_sync.py.disabled
Comment (sinmi-hub) [reply tests/assets/slack_client/test_message_sync.py.disabled:1]: Test has been refactored back to the original state and solved. It is now passing

**Comment by sinmi-hub** [comment]
Current status of security scans - One failing test

The tests are passing locally, but failing within the CI/CD environment. This is due to the moto library used in tests/assets/slack_events_bolt_handler/test_slack_events_bolt_handler.py

### Problem Summary
Root Cause: The moto library internally uses the responses library to mock HTTP requests, which conflicts with the requests library used directly in the codebase.

### Environment Difference:
Local: Tests pass because the environment has better isolation or different dependency versions
GitHub Actions CI: Tests fail due to the moto/responses/requests library conflict

Unsure how to proceed

**Review by nshalabh — APPROVED** [review]
